# Advanced Functions in Pandas

In [ ]:
%pip install pandas

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv("https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv")

In [4]:
pd.read_csv("https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv", header= None)

0         1       2   \
0    PassengerId  Survived  Pclass   
1              1         0       3   
2              2         1       1   
3              3         1       3   
4              4         1       1   
..           ...       ...     ...   
887          887         0       2   
888          888         1       1   
889          889         0       3   
890          890         1       1   
891          891         0       3   

                                                    3       4    5      6   \
0                                                 Name     Sex  Age  SibSp   
1                              Braund, Mr. Owen Harris    male   22      1   
2    Cumings, Mrs. John Bradley (Florence Briggs Th...  female   38      1   
3                                Heikkinen, Miss Laina  female   26      0   
4         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   35      1   
..                                                 ...     ...  ...    ...   
887                              Montvila, Rev. Juozas    male   27      0   
888                        Graham, Miss Margaret Edith  female   19      0   
889            Johnston, Miss Catherine Helen "Carrie"  female  NaN      1   
890                              Behr, Mr. Karl Howell    male   26      0   
891                                Dooley, Mr. Patrick    male   32      0   

        7                 8        9      10        11  
0    Parch            Ticket     Fare  Cabin  Embarked  
1        0         A/5 21171     7.25    NaN         S  
2        0          PC 17599  71.2833    C85         C  
3        0  STON/O2. 3101282    7.925    NaN         S  
4        0            113803     53.1   C123         S  
..     ...               ...      ...    ...       ...  
887      0            211536       13    NaN         S  
888      0            112053       30    B42         S  
889      2        W./C. 6607    23.45    NaN         S  
890      0            111369       30   C148         C  
891      0            370376     7.75    NaN         Q  

[892 rows x 12 columns]

# Accessing Row and Column of DataFrame

## Accessing 1st nth row

`dataframe[x:y:z]`

- x : Start , y : End, z = Steps
- Just like string slicing

In [5]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                                Heikkinen, Miss Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                        Graham, Miss Margaret Edith  female  19.0      0   
888            Johnston, Miss Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [6]:
df[0:100]

PassengerId  Survived  Pclass  \
0             1         0       3   
1             2         1       1   
2             3         1       3   
3             4         1       1   
4             5         0       3   
..          ...       ...     ...   
95           96         0       3   
96           97         0       1   
97           98         1       1   
98           99         1       2   
99          100         0       2   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss Laina  female  26.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
..                                                ...     ...   ...    ...   
95                        Shorney, Mr. Charles Joseph    male   NaN      0   
96                          Goldschmidt, Mr. George B    male  71.0      0   
97                    Greenfield, Mr. William Bertram    male  23.0      0   
98               Doling, Mrs. John T (Ada Julia Bone)  female  34.0      0   
99                                  Kantor, Mr. Sinai    male  34.0      1   

    Parch            Ticket     Fare    Cabin Embarked  
0       0         A/5 21171   7.2500      NaN        S  
1       0          PC 17599  71.2833      C85        C  
2       0  STON/O2. 3101282   7.9250      NaN        S  
3       0            113803  53.1000     C123        S  
4       0            373450   8.0500      NaN        S  
..    ...               ...      ...      ...      ...  
95      0            374910   8.0500      NaN        S  
96      0          PC 17754  34.6542       A5        C  
97      1          PC 17759  63.3583  D10 D12        C  
98      1            231919  23.0000      NaN        S  
99      0            244367  26.0000      NaN        S  

[100 rows x 12 columns]

In [7]:
df[0:100:5]

PassengerId  Survived  Pclass  \
0             1         0       3   
5             6         0       3   
10           11         1       3   
15           16         1       2   
20           21         0       2   
25           26         1       3   
30           31         0       1   
35           36         0       1   
40           41         0       3   
45           46         0       3   
50           51         0       3   
55           56         1       1   
60           61         0       3   
65           66         1       3   
70           71         0       2   
75           76         0       3   
80           81         0       3   
85           86         1       3   
90           91         0       3   
95           96         0       3   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
5                                    Moran, Mr. James    male   NaN      0   
10                     Sandstrom, Miss Marguerite Rut  female   4.0      1   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
20                               Fynney, Mr. Joseph J    male  35.0      0   
25  Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.0      1   
30                           Uruchurtu, Don. Manuel E    male  40.0      0   
35                     Holverson, Mr. Alexander Oskar    male  42.0      1   
40     Ahlin, Mrs. Johan (Johanna Persdotter Larsson)  female  40.0      1   
45                           Rogers, Mr. William John    male   NaN      0   
50                          Panula, Master Juha Niilo    male   7.0      4   
55                                  Woolner, Mr. Hugh    male   NaN      0   
60                              Sirayanian, Mr. Orsen    male  22.0      0   
65                            Moubarek, Master Gerios    male   NaN      1   
70                         Jenkin, Mr. Stephen Curnow    male  32.0      0   
75                            Moen, Mr. Sigurd Hansen    male  25.0      0   
80                               Waelens, Mr. Achille    male  22.0      0   
85  Backstrom, Mrs. Karl Alfred (Maria Mathilda Gu...  female  33.0      3   
90                               Christmann, Mr. Emil    male  29.0      0   
95                        Shorney, Mr. Charles Joseph    male   NaN      0   

    Parch           Ticket     Fare  Cabin Embarked  
0       0        A/5 21171   7.2500    NaN        S  
5       0           330877   8.4583    NaN        Q  
10      1          PP 9549  16.7000     G6        S  
15      0           248706  16.0000    NaN        S  
20      0           239865  26.0000    NaN        S  
25      5           347077  31.3875    NaN        S  
30      0         PC 17601  27.7208    NaN        C  
35      0           113789  52.0000    NaN        S  
40      0             7546   9.4750    NaN        S  
45      0  S.C./A.4. 23567   8.0500    NaN        S  
50      1          3101295  39.6875    NaN        S  
55      0            19947  35.5000    C52        S  
60      0             2669   7.2292    NaN        C  
65      1             2661  15.2458    NaN        C  
70      0       C.A. 33111  10.5000    NaN        S  
75      0           348123   7.6500  F G73        S  
80      0           345767   9.0000    NaN        S  
85      0          3101278  15.8500    NaN        S  
90      0           343276   8.0500    NaN        S  
95      0           374910   8.0500    NaN        S

## Implicit / Integer / Internal Indexing

iloc = implicit , integer or internal indexing

- Always require python internal indexing
- whatever indexing Pandas give by-default , we can use it with 

In [8]:
df[0:3]

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [9]:
df.iloc[0:3]  # start from 0 and go till 2 (3-1)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

## Explicit / Name Indexing

These indexing in which end point is also included.

In [10]:
df.loc[0:3]

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S

In [11]:
df.loc[0:2, ["Name", "Sex"]]  # if we use iloc in place of loc, we will get error as iloc in number indexing

Name     Sex
0                            Braund, Mr. Owen Harris    male
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female
2                              Heikkinen, Miss Laina  female

If we want to use iloc only, we have to use indexing in place of name.
- like in place of col name "Sex" we will use index no. 4

In [12]:
df.iloc[0:3, 3:5]

Name     Sex
0                            Braund, Mr. Owen Harris    male
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female
2                              Heikkinen, Miss Laina  female

In [13]:
df["Name"]

0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                  Heikkinen, Miss Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
                             ...                        
886                                Montvila, Rev. Juozas
887                          Graham, Miss Margaret Edith
888              Johnston, Miss Catherine Helen "Carrie"
889                                Behr, Mr. Karl Howell
890                                  Dooley, Mr. Patrick
Name: Name, Length: 891, dtype: object

In [14]:
list(df["Name"][2:5])

['Heikkinen, Miss Laina',
 'Futrelle, Mrs. Jacques Heath (Lily May Peel)',
 'Allen, Mr. William Henry']

Converting List to series

In [15]:
pd.Series(list(df["Name"][2:5]))

0                           Heikkinen, Miss Laina
1    Futrelle, Mrs. Jacques Heath (Lily May Peel)
2                        Allen, Mr. William Henry
dtype: object

In [16]:
s = pd.Series(list(df["Name"][2:5]), index = ["a", "b", "c"])
s

a                           Heikkinen, Miss Laina
b    Futrelle, Mrs. Jacques Heath (Lily May Peel)
c                        Allen, Mr. William Henry
dtype: object

In [17]:
s1 = pd.Series(list(df["Name"][5:8]))
s1

0                 Moran, Mr. James
1          McCarthy, Mr. Timothy J
2    Palsson, Master Gosta Leonard
dtype: object

### Concatenate two string

Series with different indices and data frame cannot be concatenated by using '+' symbol


series s has index [a, b, c]


series s1 has index [0, 1, 2]

In [18]:
s+s1

a    NaN
b    NaN
c    NaN
0    NaN
1    NaN
2    NaN
dtype: object

So we have to keep index same for series.

In [19]:
s1 = pd.Series(list(df["Name"][5:8]), index = ["a", "b", "c"])
s1

a                 Moran, Mr. James
b          McCarthy, Mr. Timothy J
c    Palsson, Master Gosta Leonard
dtype: object

In [20]:
s

a                           Heikkinen, Miss Laina
b    Futrelle, Mrs. Jacques Heath (Lily May Peel)
c                        Allen, Mr. William Henry
dtype: object

In [21]:
s+s1

a                Heikkinen, Miss LainaMoran, Mr. James
b    Futrelle, Mrs. Jacques Heath (Lily May Peel)Mc...
c    Allen, Mr. William HenryPalsson, Master Gosta ...
dtype: object

### Dropping any column from Data Frame

Whatever keys we are dropping, by default it is assuming in the row.
- We have to drop by giving parameter axis=1

- By default not modified ni original data frame.
- In order to modify from original data frame, we pass a parameter inplace= True

In [22]:
# df.drop('PassengerId') # error

In [23]:
df.drop("PassengerId", axis=1)

Survived  Pclass                                               Name  \
0           0       3                            Braund, Mr. Owen Harris   
1           1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2           1       3                              Heikkinen, Miss Laina   
3           1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4           0       3                           Allen, Mr. William Henry   
..        ...     ...                                                ...   
886         0       2                              Montvila, Rev. Juozas   
887         1       1                        Graham, Miss Margaret Edith   
888         0       3            Johnston, Miss Catherine Helen "Carrie"   
889         1       1                              Behr, Mr. Karl Howell   
890         0       3                                Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S  
1    female  38.0      1      0          PC 17599  71.2833   C85        C  
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
3    female  35.0      1      0            113803  53.1000  C123        S  
4      male  35.0      0      0            373450   8.0500   NaN        S  
..      ...   ...    ...    ...               ...      ...   ...      ...  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

In [24]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                                Heikkinen, Miss Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                        Graham, Miss Margaret Edith  female  19.0      0   
888            Johnston, Miss Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [25]:
df.drop("PassengerId", axis=1, inplace=True)

In [26]:
df

Survived  Pclass                                               Name  \
0           0       3                            Braund, Mr. Owen Harris   
1           1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2           1       3                              Heikkinen, Miss Laina   
3           1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4           0       3                           Allen, Mr. William Henry   
..        ...     ...                                                ...   
886         0       2                              Montvila, Rev. Juozas   
887         1       1                        Graham, Miss Margaret Edith   
888         0       3            Johnston, Miss Catherine Helen "Carrie"   
889         1       1                              Behr, Mr. Karl Howell   
890         0       3                                Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S  
1    female  38.0      1      0          PC 17599  71.2833   C85        C  
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
3    female  35.0      1      0            113803  53.1000  C123        S  
4      male  35.0      0      0            373450   8.0500   NaN        S  
..      ...   ...    ...    ...               ...      ...   ...      ...  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

Dropping the row, just mention the index of the row to drop, and inplace=True if want to drop from original data frame

In [27]:
df.drop(1, inplace=True) 

In [28]:
df

Survived  Pclass                                          Name     Sex  \
0           0       3                       Braund, Mr. Owen Harris    male   
2           1       3                         Heikkinen, Miss Laina  female   
3           1       1  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4           0       3                      Allen, Mr. William Henry    male   
5           0       3                              Moran, Mr. James    male   
..        ...     ...                                           ...     ...   
886         0       2                         Montvila, Rev. Juozas    male   
887         1       1                   Graham, Miss Margaret Edith  female   
888         0       3       Johnston, Miss Catherine Helen "Carrie"  female   
889         1       1                         Behr, Mr. Karl Howell    male   
890         0       3                           Dooley, Mr. Patrick    male   

      Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0    22.0      1      0         A/5 21171   7.2500   NaN        S  
2    26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
3    35.0      1      0            113803  53.1000  C123        S  
4    35.0      0      0            373450   8.0500   NaN        S  
5     NaN      0      0            330877   8.4583   NaN        Q  
..    ...    ...    ...               ...      ...   ...      ...  
886  27.0      0      0            211536  13.0000   NaN        S  
887  19.0      0      0            112053  30.0000   B42        S  
888   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889  26.0      0      0            111369  30.0000  C148        C  
890  32.0      0      0            370376   7.7500   NaN        Q  

[890 rows x 11 columns]

In [29]:
df.drop(4, inplace=True)

In [30]:
df

Survived  Pclass                                          Name     Sex  \
0           0       3                       Braund, Mr. Owen Harris    male   
2           1       3                         Heikkinen, Miss Laina  female   
3           1       1  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
5           0       3                              Moran, Mr. James    male   
6           0       1                       McCarthy, Mr. Timothy J    male   
..        ...     ...                                           ...     ...   
886         0       2                         Montvila, Rev. Juozas    male   
887         1       1                   Graham, Miss Margaret Edith  female   
888         0       3       Johnston, Miss Catherine Helen "Carrie"  female   
889         1       1                         Behr, Mr. Karl Howell    male   
890         0       3                           Dooley, Mr. Patrick    male   

      Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0    22.0      1      0         A/5 21171   7.2500   NaN        S  
2    26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
3    35.0      1      0            113803  53.1000  C123        S  
5     NaN      0      0            330877   8.4583   NaN        Q  
6    54.0      0      0             17463  51.8625   E46        S  
..    ...    ...    ...               ...      ...   ...      ...  
886  27.0      0      0            211536  13.0000   NaN        S  
887  19.0      0      0            112053  30.0000   B42        S  
888   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889  26.0      0      0            111369  30.0000  C148        C  
890  32.0      0      0            370376   7.7500   NaN        Q  

[889 rows x 11 columns]

To see original data frame after dropping

In [31]:
df.reset_index(drop=True)

Survived  Pclass                                          Name     Sex  \
0           0       3                       Braund, Mr. Owen Harris    male   
1           1       3                         Heikkinen, Miss Laina  female   
2           1       1  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
3           0       3                              Moran, Mr. James    male   
4           0       1                       McCarthy, Mr. Timothy J    male   
..        ...     ...                                           ...     ...   
884         0       2                         Montvila, Rev. Juozas    male   
885         1       1                   Graham, Miss Margaret Edith  female   
886         0       3       Johnston, Miss Catherine Helen "Carrie"  female   
887         1       1                         Behr, Mr. Karl Howell    male   
888         0       3                           Dooley, Mr. Patrick    male   

      Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0    22.0      1      0         A/5 21171   7.2500   NaN        S  
1    26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
2    35.0      1      0            113803  53.1000  C123        S  
3     NaN      0      0            330877   8.4583   NaN        Q  
4    54.0      0      0             17463  51.8625   E46        S  
..    ...    ...    ...               ...      ...   ...      ...  
884  27.0      0      0            211536  13.0000   NaN        S  
885  19.0      0      0            112053  30.0000   B42        S  
886   NaN      1      2        W./C. 6607  23.4500   NaN        S  
887  26.0      0      0            111369  30.0000  C148        C  
888  32.0      0      0            370376   7.7500   NaN        Q  

[889 rows x 11 columns]

### Setting a New Index

In [32]:
df.set_index("Name", inplace=True)

In [33]:
df

Survived  Pclass     Sex   Age  \
Name                                                                           
Braund, Mr. Owen Harris                              0       3    male  22.0   
Heikkinen, Miss Laina                                1       3  female  26.0   
Futrelle, Mrs. Jacques Heath (Lily May Peel)         1       1  female  35.0   
Moran, Mr. James                                     0       3    male   NaN   
McCarthy, Mr. Timothy J                              0       1    male  54.0   
...                                                ...     ...     ...   ...   
Montvila, Rev. Juozas                                0       2    male  27.0   
Graham, Miss Margaret Edith                          1       1  female  19.0   
Johnston, Miss Catherine Helen "Carrie"              0       3  female   NaN   
Behr, Mr. Karl Howell                                1       1    male  26.0   
Dooley, Mr. Patrick                                  0       3    male  32.0   

                                              SibSp  Parch            Ticket  \
Name                                                                           
Braund, Mr. Owen Harris                           1      0         A/5 21171   
Heikkinen, Miss Laina                             0      0  STON/O2. 3101282   
Futrelle, Mrs. Jacques Heath (Lily May Peel)      1      0            113803   
Moran, Mr. James                                  0      0            330877   
McCarthy, Mr. Timothy J                           0      0             17463   
...                                             ...    ...               ...   
Montvila, Rev. Juozas                             0      0            211536   
Graham, Miss Margaret Edith                       0      0            112053   
Johnston, Miss Catherine Helen "Carrie"           1      2        W./C. 6607   
Behr, Mr. Karl Howell                             0      0            111369   
Dooley, Mr. Patrick                               0      0            370376   

                                                 Fare Cabin Embarked  
Name                                                                  
Braund, Mr. Owen Harris                        7.2500   NaN        S  
Heikkinen, Miss Laina                          7.9250   NaN        S  
Futrelle, Mrs. Jacques Heath (Lily May Peel)  53.1000  C123        S  
Moran, Mr. James                               8.4583   NaN        Q  
McCarthy, Mr. Timothy J                       51.8625   E46        S  
...                                               ...   ...      ...  
Montvila, Rev. Juozas                         13.0000   NaN        S  
Graham, Miss Margaret Edith                   30.0000   B42        S  
Johnston, Miss Catherine Helen "Carrie"       23.4500   NaN        S  
Behr, Mr. Karl Howell                         30.0000  C148        C  
Dooley, Mr. Patrick                            7.7500   NaN        Q  

[889 rows x 10 columns]

Setting the index back to org form

In [34]:
df.reset_index()

Name  Survived  Pclass     Sex  \
0                         Braund, Mr. Owen Harris         0       3    male   
1                           Heikkinen, Miss Laina         1       3  female   
2    Futrelle, Mrs. Jacques Heath (Lily May Peel)         1       1  female   
3                                Moran, Mr. James         0       3    male   
4                         McCarthy, Mr. Timothy J         0       1    male   
..                                            ...       ...     ...     ...   
884                         Montvila, Rev. Juozas         0       2    male   
885                   Graham, Miss Margaret Edith         1       1  female   
886       Johnston, Miss Catherine Helen "Carrie"         0       3  female   
887                         Behr, Mr. Karl Howell         1       1    male   
888                           Dooley, Mr. Patrick         0       3    male   

      Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0    22.0      1      0         A/5 21171   7.2500   NaN        S  
1    26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
2    35.0      1      0            113803  53.1000  C123        S  
3     NaN      0      0            330877   8.4583   NaN        Q  
4    54.0      0      0             17463  51.8625   E46        S  
..    ...    ...    ...               ...      ...   ...      ...  
884  27.0      0      0            211536  13.0000   NaN        S  
885  19.0      0      0            112053  30.0000   B42        S  
886   NaN      1      2        W./C. 6607  23.4500   NaN        S  
887  26.0      0      0            111369  30.0000  C148        C  
888  32.0      0      0            370376   7.7500   NaN        Q  

[889 rows x 11 columns]

Getting Details of any particular Key from Data frame

In [35]:
df.loc['Braund, Mr. Owen Harris']

Survived            0
Pclass              3
Sex              male
Age              22.0
SibSp               1
Parch               0
Ticket      A/5 21171
Fare             7.25
Cabin             NaN
Embarked            S
Name: Braund, Mr. Owen Harris, dtype: object

### Converting Dictionary into DataFrame

In [36]:
# Key Values are in list
d = {'Key1': [2, 3, 4],
     'Key2': [5, 6, 7],
     'Key3': [8, 9, 1]}
d

{'Key1': [2, 3, 4], 'Key2': [5, 6, 7], 'Key3': [8, 9, 1]}

In [37]:
pd.DataFrame(d)

Key1  Key2  Key3
0     2     5     8
1     3     6     9
2     4     7     1

In [38]:
# Key Values are in tuple
d = {'Key1': (2, 3, 4),
     'Key2': (5, 6, 7),
     'Key3': (8, 9, 1)}
d

{'Key1': (2, 3, 4), 'Key2': (5, 6, 7), 'Key3': (8, 9, 1)}

In [39]:
pd.DataFrame(d)

Key1  Key2  Key3
0     2     5     8
1     3     6     9
2     4     7     1

In [ ]:
df1 = pd.read_csv('taxonomy.csv')

In [ ]:
df1

taxonomy_id                          name parent_id           parent_name
0           101                     Emergency       NaN                   NaN
1        101-01             Disaster Response       101             Emergency
2        101-02                Emergency Cash       101             Emergency
3     101-02-01             Help Pay for Food    101-02        Emergency Cash
4     101-02-02       Help Pay for Healthcare    101-02        Emergency Cash
..          ...                           ...       ...                   ...
285   111-01-07              Workplace Rights    111-01  Advocacy & Legal Aid
286      111-02                     Mediation       111                 Legal
287      111-03                        Notary       111                 Legal
288      111-04                Representation       111                 Legal
289      111-05  Translation & Interpretation       111                 Legal

[290 rows x 4 columns]

In [ ]:
df1.shape

(290, 4)

In [ ]:
df1.describe()

taxonomy_id          name  parent_id       parent_name
count          290           290        279               279
unique         290           183         60                50
top            101  Nursing Home  106-06-07  Health Education
freq             1             4         11                15

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 290 entries, 0 to 289
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   taxonomy_id  290 non-null    object
 1   name         290 non-null    object
 2   parent_id    279 non-null    object
 3   parent_name  279 non-null    object
dtypes: object(4)
memory usage: 9.2+ KB


### Null Values in Data Frame

In [ ]:
df1.isnull().sum()

taxonomy_id     0
name            0
parent_id      11
parent_name    11
dtype: int64

In ML we have to remove the Null Value
- Either delete row and col or Imputation

`Imputation` : Replacing Null Value with Some Value

- Numerical data will be replaced by mean, median.
- Categorical Data will be replaced by Mode.


Removing Null Value from all the rows

In [ ]:
df1.shape  # 290 rows

(290, 4)

`dataframe.dropna()` by default axis = 0 , so remove null value from rows only.
- In order to remove null value from col, just give axis =1 

In [ ]:
df1.dropna()  # remove null values from rows

taxonomy_id                          name parent_id           parent_name
1        101-01             Disaster Response       101             Emergency
2        101-02                Emergency Cash       101             Emergency
3     101-02-01             Help Pay for Food    101-02        Emergency Cash
4     101-02-02       Help Pay for Healthcare    101-02        Emergency Cash
5     101-02-03          Help Pay for Housing    101-02        Emergency Cash
..          ...                           ...       ...                   ...
285   111-01-07              Workplace Rights    111-01  Advocacy & Legal Aid
286      111-02                     Mediation       111                 Legal
287      111-03                        Notary       111                 Legal
288      111-04                Representation       111                 Legal
289      111-05  Translation & Interpretation       111                 Legal

[279 rows x 4 columns]

Here we have 279 rows.

Removing null values from col

In [ ]:
df1.dropna(axis= 1)  # remove null values from cols

taxonomy_id                          name
0           101                     Emergency
1        101-01             Disaster Response
2        101-02                Emergency Cash
3     101-02-01             Help Pay for Food
4     101-02-02       Help Pay for Healthcare
..          ...                           ...
285   111-01-07              Workplace Rights
286      111-02                     Mediation
287      111-03                        Notary
288      111-04                Representation
289      111-05  Translation & Interpretation

[290 rows x 2 columns]

In [ ]:
df1

taxonomy_id                          name parent_id           parent_name
0           101                     Emergency       NaN                   NaN
1        101-01             Disaster Response       101             Emergency
2        101-02                Emergency Cash       101             Emergency
3     101-02-01             Help Pay for Food    101-02        Emergency Cash
4     101-02-02       Help Pay for Healthcare    101-02        Emergency Cash
..          ...                           ...       ...                   ...
285   111-01-07              Workplace Rights    111-01  Advocacy & Legal Aid
286      111-02                     Mediation       111                 Legal
287      111-03                        Notary       111                 Legal
288      111-04                Representation       111                 Legal
289      111-05  Translation & Interpretation       111                 Legal

[290 rows x 4 columns]

Dropping Null value from any particular column

In [ ]:
df1[["name"]].dropna(axis=1)  # drop from one col

name
0                       Emergency
1               Disaster Response
2                  Emergency Cash
3               Help Pay for Food
4         Help Pay for Healthcare
..                            ...
285              Workplace Rights
286                     Mediation
287                        Notary
288                Representation
289  Translation & Interpretation

[290 rows x 1 columns]

### Imputation 

Replacing missing value with some value.
- Numeric data will be replaced by mean and median >>> Outlier(abnormal) value treatment is done , then use mean. otherwise median

- Categorical data will be replaced by Mode.
- Input missing value with someconstan

In [ ]:
df1.fillna("SomeValue")

taxonomy_id                          name  parent_id           parent_name
0           101                     Emergency  SomeValue             SomeValue
1        101-01             Disaster Response        101             Emergency
2        101-02                Emergency Cash        101             Emergency
3     101-02-01             Help Pay for Food     101-02        Emergency Cash
4     101-02-02       Help Pay for Healthcare     101-02        Emergency Cash
..          ...                           ...        ...                   ...
285   111-01-07              Workplace Rights     111-01  Advocacy & Legal Aid
286      111-02                     Mediation        111                 Legal
287      111-03                        Notary        111                 Legal
288      111-04                Representation        111                 Legal
289      111-05  Translation & Interpretation        111                 Legal

[290 rows x 4 columns]

Since parent_id is numerical col, we can replace NAN by 0 as

In [ ]:
df1.fillna(0)

taxonomy_id                          name parent_id           parent_name
0           101                     Emergency         0                     0
1        101-01             Disaster Response       101             Emergency
2        101-02                Emergency Cash       101             Emergency
3     101-02-01             Help Pay for Food    101-02        Emergency Cash
4     101-02-02       Help Pay for Healthcare    101-02        Emergency Cash
..          ...                           ...       ...                   ...
285   111-01-07              Workplace Rights    111-01  Advocacy & Legal Aid
286      111-02                     Mediation       111                 Legal
287      111-03                        Notary       111                 Legal
288      111-04                Representation       111                 Legal
289      111-05  Translation & Interpretation       111                 Legal

[290 rows x 4 columns]

In [ ]:
data1 = {
    'A': [1, 2, None, 4, 5, None, 7, 8, 9, 10],
    'B': [None, 11, 12, 13, None, 15, 16, None, 18, 19]
}
df2 = pd.DataFrame(data1)

In [ ]:
df2

A     B
0   1.0   NaN
1   2.0  11.0
2   NaN  12.0
3   4.0  13.0
4   5.0   NaN
5   NaN  15.0
6   7.0  16.0
7   8.0   NaN
8   9.0  18.0
9  10.0  19.0

Calculating Mean of Col

In [ ]:
df2['A'].mean()

np.float64(5.75)

Missing value is imputed by mean oc 

In [ ]:
df2['A'].fillna(df2['A'].mean())

0     1.00
1     2.00
2     5.75
3     4.00
4     5.00
5     5.75
6     7.00
7     8.00
8     9.00
9    10.00
Name: A, dtype: float64

In [ ]:
df2['A'].median()

np.float64(6.0)

In [ ]:
df2['A'].fillna(df2['A'].median())

0     1.0
1     2.0
2     6.0
3     4.0
4     5.0
5     6.0
6     7.0
7     8.0
8     9.0
9    10.0
Name: A, dtype: float64

In [ ]:
df2.fillna(0)

A     B
0   1.0   0.0
1   2.0  11.0
2   0.0  12.0
3   4.0  13.0
4   5.0   0.0
5   0.0  15.0
6   7.0  16.0
7   8.0   0.0
8   9.0  18.0
9  10.0  19.0

### Other way to impute missing value

`f-fill` : Forward filling

See the data from the bottom and fill the previous observed value to NaN value

In [ ]:
df2.ffill()

A     B
0   1.0   NaN
1   2.0  11.0
2   2.0  12.0
3   4.0  13.0
4   5.0  13.0
5   5.0  15.0
6   7.0  16.0
7   8.0  16.0
8   9.0  18.0
9  10.0  19.0

`bfill` : Backward Filling

See the data frame from top and fill the next observed value

In [ ]:
df2.bfill()

A     B
0   1.0  11.0
1   2.0  11.0
2   4.0  12.0
3   4.0  13.0
4   5.0  15.0
5   7.0  15.0
6   7.0  16.0
7   8.0  18.0
8   9.0  18.0
9  10.0  19.0

### Chechking duplicate rows

In [ ]:
df2.duplicated().sum()

np.int64(0)

In [ ]:
df

Survived  Pclass     Sex   Age  \
Name                                                                           
Braund, Mr. Owen Harris                              0       3    male  22.0   
Heikkinen, Miss Laina                                1       3  female  26.0   
Futrelle, Mrs. Jacques Heath (Lily May Peel)         1       1  female  35.0   
Moran, Mr. James                                     0       3    male   NaN   
McCarthy, Mr. Timothy J                              0       1    male  54.0   
...                                                ...     ...     ...   ...   
Montvila, Rev. Juozas                                0       2    male  27.0   
Graham, Miss Margaret Edith                          1       1  female  19.0   
Johnston, Miss Catherine Helen "Carrie"              0       3  female   NaN   
Behr, Mr. Karl Howell                                1       1    male  26.0   
Dooley, Mr. Patrick                                  0       3    male  32.0   

                                              SibSp  Parch            Ticket  \
Name                                                                           
Braund, Mr. Owen Harris                           1      0         A/5 21171   
Heikkinen, Miss Laina                             0      0  STON/O2. 3101282   
Futrelle, Mrs. Jacques Heath (Lily May Peel)      1      0            113803   
Moran, Mr. James                                  0      0            330877   
McCarthy, Mr. Timothy J                           0      0             17463   
...                                             ...    ...               ...   
Montvila, Rev. Juozas                             0      0            211536   
Graham, Miss Margaret Edith                       0      0            112053   
Johnston, Miss Catherine Helen "Carrie"           1      2        W./C. 6607   
Behr, Mr. Karl Howell                             0      0            111369   
Dooley, Mr. Patrick                               0      0            370376   

                                                 Fare Cabin Embarked  
Name                                                                  
Braund, Mr. Owen Harris                        7.2500   NaN        S  
Heikkinen, Miss Laina                          7.9250   NaN        S  
Futrelle, Mrs. Jacques Heath (Lily May Peel)  53.1000  C123        S  
Moran, Mr. James                               8.4583   NaN        Q  
McCarthy, Mr. Timothy J                       51.8625   E46        S  
...                                               ...   ...      ...  
Montvila, Rev. Juozas                         13.0000   NaN        S  
Graham, Miss Margaret Edith                   30.0000   B42        S  
Johnston, Miss Catherine Helen "Carrie"       23.4500   NaN        S  
Behr, Mr. Karl Howell                         30.0000  C148        C  
Dooley, Mr. Patrick                            7.7500   NaN        Q  

[889 rows x 10 columns]

To describe any one column we have

In [ ]:
df.Age.describe()

count    712.000000
mean      29.680014
std       14.542211
min        0.420000
25%       20.000000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

In [ ]:
df.Fare.describe()

count    889.000000
mean      32.187420
std       49.725462
min        0.000000
25%        7.895800
50%       14.454200
75%       31.000000
max      512.329200
Name: Fare, dtype: float64

In [ ]:
df.reset_index(inplace=True)

In [ ]:
df

Name  Survived  Pclass     Sex  \
0                         Braund, Mr. Owen Harris         0       3    male   
1                           Heikkinen, Miss Laina         1       3  female   
2    Futrelle, Mrs. Jacques Heath (Lily May Peel)         1       1  female   
3                                Moran, Mr. James         0       3    male   
4                         McCarthy, Mr. Timothy J         0       1    male   
..                                            ...       ...     ...     ...   
884                         Montvila, Rev. Juozas         0       2    male   
885                   Graham, Miss Margaret Edith         1       1  female   
886       Johnston, Miss Catherine Helen "Carrie"         0       3  female   
887                         Behr, Mr. Karl Howell         1       1    male   
888                           Dooley, Mr. Patrick         0       3    male   

      Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0    22.0      1      0         A/5 21171   7.2500   NaN        S  
1    26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
2    35.0      1      0            113803  53.1000  C123        S  
3     NaN      0      0            330877   8.4583   NaN        Q  
4    54.0      0      0             17463  51.8625   E46        S  
..    ...    ...    ...               ...      ...   ...      ...  
884  27.0      0      0            211536  13.0000   NaN        S  
885  19.0      0      0            112053  30.0000   B42        S  
886   NaN      1      2        W./C. 6607  23.4500   NaN        S  
887  26.0      0      0            111369  30.0000  C148        C  
888  32.0      0      0            370376   7.7500   NaN        Q  

[889 rows x 11 columns]

### Data Frame based question using Pandas

1. What is the average fare paid by the people who survived?

In [ ]:
df.Fare.mean()

np.float64(32.18741957255343)

In [ ]:
df[df["Survived"] == 1]

Name  Survived  Pclass  \
1                                Heikkinen, Miss Laina         1       3   
2         Futrelle, Mrs. Jacques Heath (Lily May Peel)         1       1   
6    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)         1       3   
7                  Nasser, Mrs. Nicholas (Adele Achem)         1       2   
8                       Sandstrom, Miss Marguerite Rut         1       3   
..                                                 ...       ...     ...   
873                    Najib, Miss Adele Kiamie "Jane"         1       3   
877      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)         1       1   
878       Shelley, Mrs. William (Imanita Parrish Hall)         1       2   
885                        Graham, Miss Margaret Edith         1       1   
887                              Behr, Mr. Karl Howell         1       1   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
1    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
2    female  35.0      1      0            113803  53.1000  C123        S  
6    female  27.0      0      2            347742  11.1333   NaN        S  
7    female  14.0      1      0            237736  30.0708   NaN        C  
8    female   4.0      1      1           PP 9549  16.7000    G6        S  
..      ...   ...    ...    ...               ...      ...   ...      ...  
873  female  15.0      0      0              2667   7.2250   NaN        C  
877  female  56.0      0      1             11767  83.1583   C50        C  
878  female  25.0      0      1            230433  26.0000   NaN        S  
885  female  19.0      0      0            112053  30.0000   B42        S  
887    male  26.0      0      0            111369  30.0000  C148        C  

[341 rows x 11 columns]

In [ ]:
df[df["Survived"] == 1]['Fare'].mean()

np.float64(48.32828768328446)

2. What is the average fare paid by the people who survived?

In [ ]:
df[df['Survived'] == 0]['Fare'].mean()

np.float64(22.143558211678833)

3. What is the average age of the people who survived?

In [ ]:
df[df['Survived'] == 1]['Age'].mean()

np.float64(28.310276816608997)

### Groupby

Grouping

![image.png](attachment:image.png)

4. Min, Max, Std of fare/age of people who survived and not survive in one go

In [ ]:
df.groupby('Survived').mean(numeric_only=True)

Pclass        Age     SibSp     Parch       Fare
Survived                                                    
0         2.531022  30.615839  0.554745  0.330292  22.143558
1         1.953079  28.310277  0.472141  0.466276  48.328288

In [ ]:
df.groupby('Survived').sum(numeric_only=True)

Pclass       Age  SibSp  Parch        Fare
Survived                                            
0           1387  12950.50    304    181  12134.6699
1            666   8181.67    161    159  16479.9461

In [ ]:
df.groupby('Survived').describe()

Pclass                                                 Age  \
          count      mean       std  min  25%  50%  75%  max  count   
Survived                                                              
0         548.0  2.531022  0.736205  1.0  2.0  3.0  3.0  3.0  423.0   
1         341.0  1.953079  0.863047  1.0  1.0  2.0  3.0  3.0  289.0   

                     ... Parch        Fare                             \
               mean  ...   75%  max  count       mean        std  min   
Survived             ...                                                
0         30.615839  ...   0.0  6.0  548.0  22.143558  31.411116  0.0   
1         28.310277  ...   1.0  5.0  341.0  48.328288  66.683277  0.0   

                                         
              25%   50%   75%       max  
Survived                                 
0          7.8542  10.5  26.0  263.0000  
1         12.4750  26.0  57.0  512.3292  

[2 rows x 40 columns]

In [ ]:
df[['Survived', 'Fare']]

Survived     Fare
0           0   7.2500
1           1   7.9250
2           1  53.1000
3           0   8.4583
4           0  51.8625
..        ...      ...
884         0  13.0000
885         1  30.0000
886         0  23.4500
887         1  30.0000
888         0   7.7500

[889 rows x 2 columns]

In [ ]:
import numpy as np

Custpmize group by

In [ ]:
df.groupby(['Survived'])['Fare'].agg(['min', 'max', 'mean', 'median', 'count', 'std', 'var'])

min       max       mean  median  count        std          var
Survived                                                                 
0         0.0  263.0000  22.143558    10.5    548  31.411116   986.658180
1         0.0  512.3292  48.328288    26.0    341  66.683277  4446.659477

Groupby based on two column

How many passengers survived/ not survived for each sex and pclass?

In [ ]:
df.groupby(['Sex', 'Pclass'])['Survived'].sum()

Sex     Pclass
female  1         90
        2         70
        3         72
male    1         45
        2         17
        3         47
Name: Survived, dtype: int64

In [ ]:
df.groupby(['Sex', 'Pclass'])['Survived'].sum().to_frame()

Survived
Sex    Pclass          
female 1             90
       2             70
       3             72
male   1             45
       2             17
       3             47

In [ ]:
df.groupby(['Sex', 'Pclass'])['Survived'].sum().to_frame().reset_index()

Sex  Pclass  Survived
0  female       1        90
1  female       2        70
2  female       3        72
3    male       1        45
4    male       2        17
5    male       3        47

In [ ]:
df.groupby(['Sex', 'Pclass'])['Survived'].sum().unstack()

Pclass   1   2   3
Sex               
female  90  70  72
male    45  17  47

In [ ]:
a = df.groupby('Pclass').sum(numeric_only=True)
a

Survived      Age  SibSp  Parch        Fare
Pclass                                             
1            135  7073.42     89     77  18106.1292
2             87  5168.83     74     70   3801.8417
3            119  8889.92    302    193   6706.6451

In [ ]:
a.T

Pclass             1          2          3
Survived    135.0000    87.0000   119.0000
Age        7073.4200  5168.8300  8889.9200
SibSp        89.0000    74.0000   302.0000
Parch        77.0000    70.0000   193.0000
Fare      18106.1292  3801.8417  6706.6451

In [ ]:
a.transpose()

Pclass             1          2          3
Survived    135.0000    87.0000   119.0000
Age        7073.4200  5168.8300  8889.9200
SibSp        89.0000    74.0000   302.0000
Parch        77.0000    70.0000   193.0000
Fare      18106.1292  3801.8417  6706.6451

In [ ]:
b = a.unstack()
b

          Pclass
Survived  1           135.0000
          2            87.0000
          3           119.0000
Age       1          7073.4200
          2          5168.8300
          3          8889.9200
SibSp     1            89.0000
          2            74.0000
          3           302.0000
Parch     1            77.0000
          2            70.0000
          3           193.0000
Fare      1         18106.1292
          2          3801.8417
          3          6706.6451
dtype: float64

In [ ]:
b.unstack()

Pclass             1          2          3
Survived    135.0000    87.0000   119.0000
Age        7073.4200  5168.8300  8889.9200
SibSp        89.0000    74.0000   302.0000
Parch        77.0000    70.0000   193.0000
Fare      18106.1292  3801.8417  6706.6451